File: src/main/java/com/koddy/server/auth/infrastructure/oauth/OAuthConnector.java
package com.koddy.server.auth.infrastructure.oauth;

import com.koddy.server.auth.domain.model.oauth.OAuthProvider;
import com.koddy.server.auth.domain.model.oauth.OAuthTokenResponse;
import com.koddy.server.auth.domain.model.oauth.OAuthUserResponse;

public interface OAuthConnector {
    boolean isSupported(final OAuthProvider provider);

    OAuthTokenResponse fetchToken(final String code, final String redirectUri, final String state);

    OAuthUserResponse fetchUserInfo(final String accessToken);

    default OAuthUserResponse login(final String code, final String redirectUri, final String state) {
        final OAuthTokenResponse token = fetchToken(code, redirectUri, state);
        return fetchUserInfo(token.accessToken());
    }

    String OAUTH_CONTENT_TYPE = "application/x-www-form-urlencoded; charset=utf-8;";

    String BEARER_TOKEN_TYPE = "Bearer";
}

File: src/main/java/com/koddy/server/auth/infrastructure/social/OAuthConnector.kt
package com.koddy.server.auth.infrastructure.social

import com.koddy.server.auth.domain.model.oauth.OAuthTokenResponse
import com.koddy.server.auth.domain.model.oauth.OAuthUserResponse

interface OAuthConnector {
    fun fetchToken(
        code: String,
        redirectUri: String,
        state: String,
    ): OAuthTokenResponse

    fun fetchUserInfo(accessToken: String): OAuthUserResponse

    fun login(
        code: String,
        redirectUri: String,
        state: String,
    ): OAuthUserResponse {
        val token: OAuthTokenResponse = fetchToken(code, redirectUri, state)
        return fetchUserInfo(token.accessToken())
    }

    companion object {
        const val OAUTH_CONTENT_TYPE: String = "application/x-www-form-urlencoded; charset=utf-8;"
        const val BEARER_TOKEN_TYPE: String = "Bearer"
    }
}
